akka-javasdk/src/main/scala/akka/javasdk/impl/reflection/Reflect.scala

```diff
@@ -160,51 +166,69 @@ private[impl] object Reflect {
-          if (parameterizedType.getActualTypeArguments.length == 1)
-            parameterizedType.getActualTypeArguments.head.asInstanceOf[Class[_]]
-          else throw new IllegalArgumentException(errorMsg)
-        case noTypeParamsParent: Class[_] =>
-          // recurse and look at parent
-          loop(noTypeParamsParent, errorMsg)
-      }
+    // register all inputs for methods returning StepEffect
+    val stepEffectClass = classOf[Workflow.StepEffect]
+    val methods = clz.getDeclaredMethods.filter { m =>
+      m.getParameterTypes.length == 1 && stepEffectClass.isAssignableFrom(m.getReturnType)
     }
 
-  def tableUpdaterRowType(tableUpdater: Class[_]): Class[_] = {
-    @tailrec
-    def loop(current: Class[_]): Class[_] =
-      if (current == classOf[AnyRef])
-        // recursed to root without finding type param
-        throw new IllegalArgumentException(s"Cannot find table updater class for ${tableUpdater.getClass}")
-      else {
-        current.getGenericSuperclass match {
-          case parameterizedType: ParameterizedType =>
-            if (parameterizedType.getActualTypeArguments.size == 1)
-              parameterizedType.getActualTypeArguments.head.asInstanceOf[Class[_]]
-            else throw new IllegalArgumentException(s"Cannot find table updater class for ${tableUpdater.getClass}")
-          case noTypeParamsParent: Class[_] =>
-            // recurse and look at parent
-            loop(noTypeParamsParent)
-        }
+    methods.foldLeft(List.empty[Class[_]]) { (acc, method) =>
+      acc ++ lookupSubClasses(method.getParameterTypes.head)
+    }
+  }
+
+  @nowarn("msg=deprecated")
+  def workflowKnownInputTypes[S, W <: Workflow[S]](workflow: Workflow[S]): List[Class[_]] =
+    workflow
+      .definition()
+      .getSteps
+      .asScala
+      .flatMap {
+        case asyncCallStep: Workflow.AsyncCallStep[_, _, _] =>
+          if (asyncCallStep.transitionInputClass == null) lookupSubClasses(asyncCallStep.callInputClass)
+          else lookupSubClasses(asyncCallStep.callInputClass) ++ lookupSubClasses(asyncCallStep.transitionInputClass)
+
+        case callStep: Workflow.CallStep[_, _, _] =>
+          if (callStep.transitionInputClass == null) lookupSubClasses(callStep.callInputClass)
+          else lookupSubClasses(callStep.callInputClass) ++ lookupSubClasses(callStep.transitionInputClass)
+
+        case runnable: RunnableStep => List.empty
+      }
+      .toList
+
+  /**
+   * This method will try to find all know subtypes if the passed class is an interface. Interfaces that are neither
+   * sealed nor have JsonSubTypes annotation cannot be handled As we can't deserialize them.
+   *
+   * Note this method is only used when registering workflow step input types
+   */
+  private def lookupSubClasses(cls: Class[_]): List[Class[_]] =
+    if (cls.isInterface) {
+      if (cls.isSealed) cls.getPermittedSubclasses.toList :+ cls
+      else if (cls.hasAnnotation[JsonSubTypes]) {
```

Comment:
I'm not sure if we must register JsonSubTypes, I think, in that case, the payload itself is enough for Jackson to deserialize it correctly. Probably it's not a mistake to register hints. 

Maybe we could verify a different approach. We could force users to always add `JsonSubTypes` for interfaces. Just thinking...